### PR TITLE
add disk for more context in bitrot errors

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
add disk for more context in bitrot errors

## Motivation and Context
adds more context for errors lower in the stack

## How to test this PR?
Nothing special product corruption if possible
,try to read the object and see that the errors
are displayed with the disk where they originated
from.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_exclamation_mark: [more...](#c1fac4f8-6c32-44eb-b38d-615923e843d3) |
| mint-fs.sh | :heavy_exclamation_mark: [more...](#024fe702-eb0f-4a9a-8b15-585ac80315a2) |
| mint-gateway-s3.sh | :heavy_exclamation_mark: [more...](#61af4ff8-072e-43f5-8b0a-69e657c1cd5c) |
| mint-gateway-azure.sh | :heavy_exclamation_mark: [more...](#8589a284-e635-4bd1-b830-35bdb2de3cd6) |
| mint-erasure.sh | :heavy_exclamation_mark: [more...](#5c9b22d5-5274-4b9d-a0c4-883912748ed4) |
| mint-dist-erasure.sh | :heavy_exclamation_mark: [more...](#95f32123-3bf7-45e4-a11a-762c4dd270b4) |
| mint-zoned.sh | :heavy_exclamation_mark: [more...](#0dddd3bb-da0b-4fc7-86f5-7464e58e6e61) |
| mint-gateway-nas.sh | :heavy_exclamation_mark: [more...](#d9f232b7-6e99-404e-af65-f8f2e88f18ad) |
<a name="0dddd3bb-da0b-4fc7-86f5-7464e58e6e61"></a>
### 10296-1251f99/mint-zoned.sh.xtrace:
```
DISABLED TIME BEING
```
<a name="95f32123-3bf7-45e4-a11a-762c4dd270b4"></a>
### 10296-1251f99/mint-dist-erasure.sh.xtrace:
```
DISABLED TIME BEING
```
<a name="5c9b22d5-5274-4b9d-a0c4-883912748ed4"></a>
### 10296-1251f99/mint-erasure.sh.xtrace:
```
DISABLED TIME BEING
```
<a name="8589a284-e635-4bd1-b830-35bdb2de3cd6"></a>
### 10296-1251f99/mint-gateway-azure.sh.xtrace:
```
DISABLED TIME BEING
```
<a name="d9f232b7-6e99-404e-af65-f8f2e88f18ad"></a>
### 10296-1251f99/mint-gateway-nas.sh.xtrace:
```
DISABLED TIME BEING
```
<a name="61af4ff8-072e-43f5-8b0a-69e657c1cd5c"></a>
### 10296-1251f99/mint-gateway-s3.sh.xtrace:
```
DISABLED TIME BEING
```
<a name="024fe702-eb0f-4a9a-8b15-585ac80315a2"></a>
### 10296-1251f99/mint-fs.sh.xtrace:
```
DISABLED TIME BEING
```
<a name="c1fac4f8-6c32-44eb-b38d-615923e843d3"></a>
### 10296-1251f99/mint-large-bucket.sh.xtrace:
```
DISABLED TIME BEING
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :x: [more...](#3edcffaa-c663-4ad1-9636-2ee3d75d2cbd) |
<a name="3edcffaa-c663-4ad1-9636-2ee3d75d2cbd"></a>
### 10296-87413b6/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:32718
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 767e26208d11:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 2 minutes and 0 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 12 seconds
(8/15) Running minio-dotnet tests ... done in 29 seconds
(9/15) Running minio-go tests ... FAILED in 24 seconds
{
  "args": {
    "destination": {
      "Bucket": "minio-go-test-9df6cskf0qkmbjca",
      "Object": "srcObject",
      "Encryption": {},
      "UserMetadata": null,
      "ReplaceMetadata": false,
      "UserTags": null,
      "ReplaceTags": false,
      "LegalHold": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "Size": 0,
      "Progress": null
    },
    "source": {
      "Bucket": "minio-go-test-9df6cskf0qkmbjca",
      "Object": "srcObject",
      "VersionID": "",
      "MatchETag": "",
      "NoMatchETag": "",
      "MatchModifiedSince": "0001-01-01T00:00:00Z",
      "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
      "MatchRange": false,
      "Start": 0,
      "End": 0,
      "Encryption": null
    }
  },
  "duration": 444,
  "error": "This copy request is illegal because it is trying to copy an object to itself without changing the object's metadata, storage class, website redirect location or encryption attributes.",
  "function": "CopyObject(destination, source)",
  "message": "CopyObject failed",
  "name": "minio-go: testEncryptedSSES3ToSSES3CopyObject",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 36 seconds
(10/15) Running minio-js tests ... done in 32 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 31 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 0 seconds
(13/15) Running s3select tests ... done in 3 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/bitrot-whole.go [hunk 1]
```diff
@@ -69,9 +70,7 @@ func (b *wholeBitrotReader) ReadAt(buf []byte, offset int64) (n int, err error)
 	if b.buf == nil {
 		b.buf = make([]byte, b.tillOffset-offset)
 		if _, err := b.disk.ReadFile(b.volume, b.filePath, offset, b.buf, b.verifier); err != nil {
-			ctx := GlobalContext
-			logger.GetReqInfo(ctx).AppendTags("disk", b.disk.String())
-			logger.LogIf(ctx, err)
+			logger.LogIf(GlobalContext, fmt.Errorf("Disk: %s return %w", b.disk, err))
```
Comment (vadmeste) [inline cmd/bitrot-whole.go]: s/return/returned

**Review by vadmeste — COMMENTED** [review]
One comment, LGTM otherwise

File: cmd/bitrot-streaming.go [hunk 2]
```diff
@@ -139,8 +139,8 @@ func (b *streamingBitrotReader) ReadAt(buf []byte, offset int64) (int, error) {
 	b.h.Write(buf)
 
 	if !bytes.Equal(b.h.Sum(nil), b.hashBytes) {
-		err := &errHashMismatch{fmt.Sprintf("hashes do not match expected %s, got %s",
-			hex.EncodeToString(b.hashBytes), hex.EncodeToString(b.h.Sum(nil)))}
+		err := &errHashMismatch{fmt.Sprintf("Disk: %s returned - hashes do not match expected %s, got %s",
```
Comment (klauspost) [inline cmd/bitrot-streaming.go]:
```suggestion
		err := &errHashMismatch{fmt.Sprintf("Disk: %s - content hash does not match - expected %s, got %s",
```

**Review by klauspost — APPROVED** [review]
Not biggie, just slightly more accurate error.

File: cmd/bitrot-whole.go (hunk 1 above)
Comment (harshavardhana) [inline cmd/bitrot-whole.go]:
```suggestion
			logger.LogIf(GlobalContext, fmt.Errorf("Disk: %s returned %w", b.disk, err))
```

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :x: [more...](#3d3aee03-15c6-44bf-a163-83107e845b1a) |
| mint-gateway-azure.sh | :x: [more...](#32cf46e5-d3da-424d-82f4-82fc36232737) |
<a name="3d3aee03-15c6-44bf-a163-83107e845b1a"></a>
### 10296-65b1113/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:30385
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e0a09d8c1f1a:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 1 minutes and 59 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 37 seconds
(8/15) Running minio-dotnet tests ... done in 31 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 6 seconds
{
  "args": {
    "destination": {
      "Bucket": "minio-go-test-cdhxvfc4t5vnzzfo",
      "Object": "srcObject",
      "Encryption": {},
      "UserMetadata": null,
      "ReplaceMetadata": false,
      "UserTags": null,
      "ReplaceTags": false,
      "LegalHold": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "Size": 0,
      "Progress": null
    },
    "source": {
      "Bucket": "minio-go-test-cdhxvfc4t5vnzzfo",
      "Object": "srcObject",
      "VersionID": "",
      "MatchETag": "",
      "NoMatchETag": "",
      "MatchModifiedSince": "0001-01-01T00:00:00Z",
      "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
      "MatchRange": false,
      "Start": 0,
      "End": 0,
      "Encryption": null
    }
  },
  "duration": 229,
  "error": "This copy request is illegal because it is trying to copy an object to itself without changing the object's metadata, storage class, website redirect location or encryption attributes.",
  "function": "CopyObject(destination, source)",
  "message": "CopyObject failed",
  "name": "minio-go: testEncryptedSSES3ToSSES3CopyObject",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 22 seconds
(10/15) Running minio-js tests ... done in 46 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 31 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 0 seconds
(13/15) Running s3select tests ... done in 3 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="32cf46e5-d3da-424d-82f4-82fc36232737"></a>
### 10296-65b1113/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c2.minio.io:31248
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d8376a8c1c6e:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 4 minutes and 38 seconds
(4/15) Running aws-sdk-ruby tests ... done in 24 seconds
(5/15) Running awscli tests ... done in 2 minutes and 43 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 2 minutes and 57 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 30 seconds
(9/15) Running minio-go tests ... FAILED in 5 minutes and 5 seconds
{
  "args": {},
  "duration": 996,
  "error": "The specified multipart upload does not exist. The upload ID may be invalid, or the upload may have been aborted or completed.",
  "function": "CopyObjectPart(destination, source)",
  "message": "CompleteMultipartUpload call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 6 minutes and 12 seconds
(10/15) Running minio-js tests ... done in 2 minutes and 37 seconds
(11/15) Running minio-py tests ... done in 10 minutes and 4 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 27 seconds
(13/15) Running s3select tests ... done in 1 minutes and 1 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally
Error: No such image: minio/minio:10296-65b1113

**Pull request merged**